rfcs/20200928-keras-cv-single-stage-2d-object-detection.md

```diff
@@ -0,0 +1,396 @@
+# keras-cv Single Stage Two-Dimensional Object Detection API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com)|
+| **Contributor(s)** | Pengchong Jin (pengchong@google.com)|
+| **Updated**   | 2020-09-28                                           |
+
+## Objective
+
+We aim at providing the core primitive components for training and serving single stage two-dimensional object
+detection models, such as Single-Shot MultiBox Detector (SSD), RetinaNet, You-Only-Look-Once (YOLO).
+
+## Key Benefits
+
+Single stage object detection models are a state-of-art technique that powers many computer vision tasks, they provide
+faster detection compared to two stage models (such as FasterRCNN), while maintaining comparable performance.
+
+With this proposal, Keras users will be able to build end-to-end models with a much simplified API.
+
+## Design overview
+
+This proposal includes the specific core components for building single stage object detection models. It does not,
+include, however, 1) Data augmentation, such as image and groundtruth box preprocessing, 2) Model backbone, such as
+DarkNet, or functions to generate feature maps 3) Detection heads, such as Feature Pyramid, 4) metrics utils such as
+COCO Evaluator, or visualization utils. Data augmentation will be included as a separate RFC that handles a
+broader context than object detection.
+
+Model backbone and detection heads are model specific, we anticipate them to be analyzed and proposed in 
+keras-applications for heavily used patterns, however the user can build them easily using keras.
+
+#### Training
+
+Case where a user want to train from scratch:
+```python
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_cv
+
+# Considering a COCO dataset
+coco_dataset = tfds.load('coco/2017')
+train_ds, eval_ds = coco_dataset['train'], coco_dataset['validation']
+
+def preprocess(features):
+  image, gt_boxes, gt_labels = features['image'], features['objects']['bbox'], features['objects']['label']
+  # preprocess image, gt_boxes, gt_labels, such as flip, resize, and padding, and reserve 0 for background label.
+  return image, gt_boxes, gt_labels
+
+anchor_generator = keras_cv.ops.AnchorGenerator(anchor_sizes, scales, aspect_ratios, strides)
```

Comment:
What are the different submodules in `keras_cv`? Everything seems to live in `ops`.

Reply:
So we proposed 2 layers, 1 loss, and 4 ops. However I'd like your input on whether `BoxCoder` and `BoxMatcher` should be layer instead, given our example workflow.